express-notes-api/server.js

```diff
@@ -0,0 +1,145 @@
+import express from 'express';
+import * as fs from 'node:fs/promises';
+
+const app = express();
+const jsonParser = express.json();
+
+app.use(jsonParser);
+
+app.use('/api/notes/:id', (req, res, next) => {
+  if (!Number(req.params.id)) {
+    res.status(400).send({ error: `id ${req.params.id} must be a positive integer.` });
+  } else {
+    next();
+  }
+});
+
+app.get('/api/notes', async (req, res) => {
+  const theObject = await read();
+  const notesList = [];
+  for (const [, value] of Object.entries(theObject.notes)) {
+    notesList.push(value);
+  }
+  res.json(notesList);
+});
+
+app.get('/api/notes/:id', async (req, res) => {
+  const theObject = await read();
+  for (const [key, value] of Object.entries(theObject.notes)) {
+    if (req.params.id === key) {
+      res.json(value);
+      return;
+    }
+  }
+  res.status(400).send(`Id ${req.params.id} not found.`);
+});
+
+app.post('/api/notes', async (req, res) => {
+  if (req.body.content) {
+    try {
+      const newNote = await create(req.body);
+      res.status(201).json(newNote);
+    } catch (err) {
+      console.error('Error occured writing to data.json:', err);
+      res.status(500).send('Something went wrong.');
+    }
+  } else {
+    res.status(400).send('Please include content to record as a note.');
+  }
+});
+
+app.delete('/api/notes/:id', async (req, res) => {
+  try {
+    const response = await deleteEntry(req.params.id);
+    if (!response) {
+      res.sendStatus(204);
+    } else {
+      res.status(404).send(response);
+    }
+  } catch (err) {
+    console.error('Theres has been an error deleting:', err);
+    res.sendStatus(500);
+  }
+});
+
+app.put('/api/notes/:id', async (req, res) => {
+  if (req.body.content) {
+    try {
+      const response = await update(req.params.id, req.body.content);
+      res.status(200).json(response);
+    } catch (err) {
+      console.error('Theres has been an error updating:', err);
+      res.sendStatus(500);
+    }
+  } else {
+    res.status(400).send('Please include content to record as a note.');
+  }
+});
+
+/**
+ * Reads and parses a JSON file.
+ * @returns An object
+ */
+async function read() {
+  try {
```

Comment:
There's a problem here with your error handling. If an error is thrown it will be caught in the `catch`, which will log it. What happens next? Execution continues and the function returns. Since no return was specified, it returns `undefined`. Then the code that called `read()` gets an unexpected `undefined` and likely crashes.

You can test this by using an illegal file name in `readFile` (like 'data.jsonx'). I think you'll see that you get the error message below followed by another unexpected error.

One way to fix it is to rethrow the error. You can rethrow an error with `throw err;`. Logging an error and rethrowing it isn't a good practice, though, since it will likely result in the error being logged twice.

The best way to fix this would be to not catch the error in this function at all. Remove the try/catch. Then if `readFile` throws, since it's not caught here, it will be caught in the caller's `catch` block.